contrib/aws/aws-sdk-go-v2/aws/aws.go

```diff
@@ -0,0 +1,134 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016 Datadog, Inc.
+
+package aws
+
+import (
+	"context"
+	"fmt"
+	"math"
+
+	awsmiddleware "github.com/aws/aws-sdk-go-v2/aws/middleware"
+	"github.com/aws/smithy-go/middleware"
+	smithyhttp "github.com/aws/smithy-go/transport/http"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/ext"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+)
+
+const (
+	tagAWSAgent     = "aws.agent"
+	tagAWSService   = "aws.service"
+	tagAWSOperation = "aws.operation"
+	tagAWSRegion    = "aws.region"
+	tagAWSRequestID = "aws.request_id"
+)
+
+// Middleware allows us to add middleware to the AWS SDK GO v2.
+// See https://aws.github.io/aws-sdk-go-v2/docs/middleware for more information.
+type Middleware struct {
+	cfg *config
+}
+
+// NewMiddleware creates a new Middleware.
+func NewMiddleware(opts ...Option) *Middleware {
+	cfg := &config{}
+
+	defaults(cfg)
+	for _, opt := range opts {
+		opt(cfg)
+	}
+
+	return &Middleware{cfg: cfg}
+}
+
+// Append takes the API options from the aws.Config.
+func (mw *Middleware) Append(apiOptions *[]func(*middleware.Stack) error) {
+	*apiOptions = append(*apiOptions, func(stack *middleware.Stack) error {
+		// Add Initialize middleware after existing AWS middleware to make sure we can get details from the context.
+		// This also allows us to start the trace as early as possible.
+		return stack.Initialize.Add(middleware.InitializeMiddlewareFunc("TraceInitialize", mw.initialize), middleware.After)
```

Comment:
Better to add dd middleware at the front to make sure the telemetry data capture the correct duration. To make sure we can capture a very rare case if user adds his custome middleware and there is a time consuming bug in it.

Reply:
I still think it's important to set the minimum metadata, such as operationName without mutation of the span (otherwise you could get a span that doesn't really mean anything). Obviously using after is the only easy way to do this.

Reply:
The embarrassing thing is caused by AWS SDK middleware, just because we don't know the service name when creating the telemetry data. It is ok that when we wrap the function we don't know anything about function at the very beginning, but not say we candnot start the trace until we have known everything :)
So, it is ok to update the span during the process.

Reply:
@gbbr I'd been keen for your thoughts here. Its cleaner to use a single middleware (like already done in this PR) and the risk is fairly minimal for any middleware to fail before our Init middleware. Someone could create some custom InitMiddleware that breaks, and if they used it before ours it wouldn't trace.

I created another branch with an extra Init middleware that runs Before - https://github.com/DataDog/dd-trace-go/compare/v1...speza:feat/%23869-contrib_aws_sdk_go_v2_other?expand=1. This would require creating a trace with bare minimum (and unknown placeholders) until we can enrich it with more metadata.